libs/api/bpf_syscall.cpp

```diff
@@ -6,188 +6,230 @@
 #include <windows.h>
 #include <WinError.h>
 #include <linux/bpf.h>
+#include <sal.h>
+#include <stdexcept>
 
-#define CHECK_SIZE(last_field_name)                                                    \
-    if (!tail_is_zero(                                                                 \
-            attr,                                                                      \
-            offsetof(union bpf_attr, last_field_name) + sizeof(attr->last_field_name), \
-            sizeof(union bpf_attr))) {                                                 \
-        return -EINVAL;                                                                \
-    }
-
-static bool
-tail_is_zero(const void* buf, unsigned int start, unsigned int end)
+template <typename T> class ExtensibleStruct
 {
-    const unsigned char* p = (const unsigned char*)buf + start;
-    const unsigned char* e = (const unsigned char*)buf + end;
+  private:
+    void* _source;
+    size_t _copy_size;
+    T _tmp;
+    T* _ptr;
 
-    for (; p < e; p++) {
-        if ((*p) != 0) {
-            return false;
+    static void
+    check_tail(_In_reads_bytes_(end) const void* buf, size_t start, size_t end)
+    {
+        const unsigned char* p = (const unsigned char*)buf + start;
+        const unsigned char* e = (const unsigned char*)buf + end;
+
+        for (; p < e; p++) {
+            if ((*p) != 0) {
+                throw std::runtime_error("Non-zero tail");
+            }
         }
     }
 
-    return true;
-}
+  public:
+    ExtensibleStruct(_In_reads_bytes_(ptr_size) void* ptr, size_t ptr_size) : _source(ptr)
```

Comment:
```suggestion
    ExtensibleStruct(_In_reads_bytes_(size) void* pointer, size_t size) : _source(pointer)
```
Unless the names need to match something outside this class.
If they're arbitrary, avoid abbreviations per coding conventions.
(note that "pointer size" is misleading since it would be confused with sizeof(void*))